Fix handlabeler/utility belt misprediction

osjarw opened the pull request:
<!-- Please read these guidelines before opening your PR: https://docs.spacestation14.io/en/getting-started/pr-guideline -->
<!-- The text between the arrows are comments - they will not be visible on your PR. -->

## About the PR
I moved `handlabelercomponent` from `Content.Server` to `Content.Shared`.

## Why / Balance
It's annoying how the hand labeler isn't predicted correctly.

## Technical details
To be honest, I don't fully understand how this is working. For example, in the second video I give the labeler to the other client, and open the UI. From what I've understood, with my current commit the labeler shouldn't retain it's label across different clients. 

## Media
<!-- 
PRs which make ingame changes (adding clothing, items, new features, etc) are required to have media attached that showcase the changes.
Small fixes/refactors are exempt.
Any media may be used in SS14 progress reports, with clear credit given.

If you're unsure whether your PR will require media, ask a maintainer.

Check the box below to confirm that you have in fact seen this (put an X in the brackets, like [X]):
-->

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

Before the fix:
I can't place the labeler in the belt via the UI, and using smart-equip fails for a second, and then a moment later puts the labeler in the belt.

https://github.com/space-wizards/space-station-14/assets/62134478/1cab4642-0f45-45a6-a2b6-79d511d6e449

After the fix:
I am able to put the labeler in the belt, and using smart-equip is instant. 
The labeler is still able to label items, as it's supposed to.


https://github.com/space-wizards/space-station-14/assets/62134478/10bd8184-c565-4af9-b8c6-896c99f02cba



## Breaking changes
`HandLabelerComponent.cs` has been moved to `Content.Shared` from `Content.Server`

**Changelog**
<!--
Make players aware of new features and changes that could affect how they play the game by adding a Changelog entry. Please read the Changelog guidelines located at: https://docs.spacestation14.io/en/getting-started/pr-guideline#changelog
-->

<!--
Make sure to take this Changelog template out of the comment block in order for it to show up.
-->
:cl:
- fix: Putting a hand labeler in a belt is now predicted correctly.

deltanedas commented on Content.Shared/Labels/Components/HandLabelerComponent.cs:
```diff
@@ -1,6 +1,7 @@
 
-namespace Content.Server.Labels.Components
+namespace Content.Shared.Labels.Components
 {
```
if you are gonna move it to shared:
1. actually predict it move the system there too and make it play nicely
2. file scope namespace
3. replace each fields attribute with just `[DataField]` the rest has been automatically done for some time
4. remove `{ get; set; }` and add access attribute for the system
5. probably add `NetworkedComponent`

osjarw replied:
Thanks for reviewing this commit, and giving concrete points to work on.
I'll get to it tomorrow, since figuring all of that out properly isn't that doable after a long day of trying to learn stuff.

osjarw commented:
Cleaned up a bit and implemented some of Deltanedas' suggestions. 
This PR should currently address what it is named after, that is, predicting the transferring of the hand labeler to the belt.

However, I think that next I'll try to properly move `HandLabelerSystem` to shared, and making it actually predict all of what it is doing.
I'll probably commit that change to this PR, if this PR is still open when I'm finished.

From what I've read so far, the most problematic parts seem to be creating the labeling UI on the client, and deciding how to handle the case of client mispredicting the act of labeling something.

The former problem is a bit difficult because I still haven't fully figured out how to properly render the labeling window, but I think I should eventually figure it out.
And I think I'll solve the latter case of mispredicting the labeling of items by spawning only 1 PopupEntity when the client succesfully predicts the label creation, and spawning another if the server disagrees with whatever the client decided.

metalgearsloth commented on Content.Shared/Labels/EntitySystems/SharedHandLabelerSystem.cs:
```diff
@@ -0,0 +1,8 @@
+namespace Content.Shared.Labels.EntitySystems;
+
+public abstract class SharedHandLabelerSystem : EntitySystem {
+
+    public override void Initialize(){
+        base.Initialize();
+    }
+}
```
Please format it like all other classes in the project.

metalgearsloth reviewed (changes requested):
System still needs moving to shared, I think some things might need to stay server but most of it can get moved.

deltanedas commented on Content.Server/Labels/Label/HandLabelerSystem.cs:
```diff
@@ -16,7 +17,7 @@ namespace Content.Server.Labels
     /// </summary>
     [UsedImplicitly]
-    public sealed class HandLabelerSystem : EntitySystem
+    public sealed class HandLabelerSystem : SharedHandLabelerSystem
```
this entire system can probably be moved to shared so labeller is predicted

osjarw commented:
Now labeling stuff is predicted, so you don't need to wait a moment to get the popup about an item being labelled. 

I don't think that I'll predict the UI creation, since that seems quite a lot harder. 
From what I can tell, there's almost no cases where opening UIs is predicted, which makes finding examples rather hard.

metalgearsloth commented on Content.Client/Labels/EntitySystems/HandLabelerSystem.cs:
```diff
@@ -0,0 +1,64 @@
+        action = LabelAction.Applied;
+        result = Loc.GetString("hand-labeler-successfully-applied");
+    }
+
+    /// <summary>
+    /// Test if our prediction matches with the server,
+    /// if it doesn't, create a popup with the valid message.
+    /// </summary>
+    private void handleMessages(HandLabelerMessage message, EntitySessionEventArgs eventArgs)
+    {
+        labeledEntities.TryGetValue(message.Target.Id, out var access);
+        labeledEntities.Remove(message.Target.Id);
+        if(access == message.Action)
+            return;
+
+        string result;
+        if (message.Action == LabelAction.Removed)
+            result = Loc.GetString("hand-labeler-successfully-removed");
+        else
+            result = Loc.GetString("hand-labeler-successfully-applied");
+        _popupSystem.PopupEntity(result, GetEntity(message.User), GetEntity(message.User));
+    }
+
```
This is not the way to handle prediction, put the code in shared and use component states.

metalgearsloth reviewed (changes requested):
1. You should never modify state in getstate by clearing recentlylabeled, this won't network properly at all.
2. You check for client in a bunch of places which will just break prediction.
3. You're applying popups in a state handler which shouldn't be done, states can be reapplied infinitely and code should handle it properly.

metalgearsloth commented:
I have rewritten this but have 1 prediction issue left before it can be merged.

osjarw commented:
> 1. You should never modify state in getstate by clearing recentlylabeled, this won't network properly at all.

Recentlylabeled wasn't really supposed to be a networked variable, it was supposed to work as a message/event informing the client of mispredictions. 
Both the client and the server should've maintained their own instance of recentlylabeled, removing it's contents as the other party sent confirmations regarding the result of the labeling. 

> 3. You're applying popups in a state handler which shouldn't be done, states can be reapplied infinitely and code should handle it properly.

What exactly is this proper way of handling it?

As to the infinite reapplying, I had already accounted for that. 
It probably didn't look like that, because I had forgotten to add one line of code as I was translating the former event based solution to a componentstate based one.

metalgearsloth commented:
The client reapplies states if the entity ever comes into range so there's going to be popups every time the client handles the state which is bad.

It's just a lot of convoluted and hard-to-maintain code trying to work around prediction when you could just use the appropriate methods and not get the bugs in the first place.

osjarw commented:
> The client reapplies states if the entity ever comes into range so there's going to be popups every time the client handles the state which is bad.

As far as I'm aware, this shouldn't be happening with this code, assuming the culling of recentlylabeled is added back, as it should be. 
Though I might be wrong, if there exists some implicit copying of the server component state to the client side, and I don't mean raising HandleState/GetState.

> It's just a lot of convoluted and hard-to-maintain code trying to work around prediction when you could just use the appropriate methods and not get the bugs in the first place.

Yeah, I must admit, my solution is rather convoluted, especially since there's a lack of comments explaining the convolution.
I ended up creating this monstrosity of a solution because I couldn't find the appropriate methods for dealing with this.
